certbot/compat/filesystem.py

```diff
@@ -36,6 +41,43 @@ def chmod(file_path, mode):
         _apply_win_mode(file_path, mode)
 
 
+def open(file_path, flags, mode=0o777):  # pylint: disable=redefined-builtin
+    # type: (str, int, int) -> int
+    """
+    Wrapper of original os.open function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int flags: Flags to apply on file while opened
+    :param int mode: POSIX mode to apply on file when opened,
+        Python defaults will be applied if ``None``
+    :returns: the file descriptor to the opened file
+    :rtype: int
+    """
+    if not POSIX_MODE:
+        # Handle creation of the file atomically with proper permissions
+        if not os.path.exists(file_path) and flags & os.O_CREAT:
+            attributes = win32security.SECURITY_ATTRIBUTES()
+            security = attributes.SECURITY_DESCRIPTOR
+            user = _get_current_user()
+            dacl = _generate_dacl(user, mode)
+            security.SetSecurityDescriptorDacl(1, dacl, 0)
+
+            handle = win32file.CreateFile(file_path, win32file.GENERIC_READ, 0,
```

Comment:
there's still a race condition here, no? the file could have been created between the `if not os.path.exists` call and here. So we don't know for sure at this line that the file doesn't exist. And we're not checking `os.O_EXCL` here, so how do we know for sure that we should error out if the file already exists? we're just letting the program crash if there's a race condition, even if they've only set `os.O_CREAT`?

Reply:
Indeed you are right. See my general comment below.